system/jlib/jutil.cpp

```diff
@@ -812,27 +817,161 @@ class CRandom: public IRandomNumberGenerator, public CInterface
         return ret;
     }
 
-} RandomMain; 
+};
 
-static CriticalSection gobalRandomSect;
+#ifdef __APPLE__
+__thread Owned<IRandomNumberGenerator> RandomMain = createRandomNumberGenerator();
+#else
+thread_local Owned<IRandomNumberGenerator> RandomMain = createRandomNumberGenerator();
```

Comment:
I don't know if/when thread-scope constructors/destructors get called, or how they interact with thread-pooling. I'm not convinced this is safe.

Even if it was, it would mean every thread that was created would create a random number generator, which is not really desirable (I don't know how expensive they are to create, maybe it's ok).

Using a per-thread random generator is probably a good idea, but I suspect it needs to be created on demand and deleted via a thread termination hook.